text/0004-import-reorg.md

```diff
@@ -0,0 +1,132 @@
+
+In other words we want to be able to say that if module A depends on module B
+it depends on the entirey of it.  Today this is violated by a `sentry.app`
+and `sentry.models`.
+
+# Motivation
+
+Sentry's CI and even local testing times are quickly increasing through the
+code size in Sentry.  We run pretty much all tests at all times as we are not
+able to determine which change is going to require which code to run.  The
+solution in parts to this will likely be to compartmentalize Sentry into smaller
+services.  Today drawing these service boundaries however is relatively tricky
+because a lot of code within Sentry is deeply intertwined.
+
+This also in parts shows up in import times.  To run a single test file for a
+utils module (`test_rust.py`) pytest will spend 0.2 seconds in test execution vs
+2 seconds in import time.  While imports are so far acceptable, the bigger issue
+caused by it is that it increases the total amount of surface that we need to
+consider for test execution.
+
+The main motivator however is the inability to draw boundaries within Sentry
+today.  For instance today `sentry.http` pulls in `sentry.models` (to import
+`EventError` to access a constant) which pulls in _all_ database models.
+`sentry.http` itself is needed by `sentry.utils.pytest.sentry` and some other
+places.  Because all models are imported, not just the model declarations are
+imported but a lot of the application.  For instance via the
+`sentry.models.identity` the `sentry.analytics` system is pulled.
+`sentry.models.integrations` will pull in the entire integration platform code
+including `sentry.pipeline` which drags in the entire incidents system, API
+helpers and more.  `sentry.models.organizationmember` pulls in `sentry.app`
+which then pulls in `tsdb`, `buffer`, `nodestore` etc.
+
+While it's undoubtedly true that today many of these imports will happen anyways
+as we are globally configuring sentry in the tests, getting imports under control
+will let us slowly break the Sentry monolith into distinct services in an easier
+and more controlled manner.
+
+Making these imports however explicit enables us to better understand the real
+dependencies through imports.  Today we cannot use import tracking to see the
+real dependencies because they are obfuscated through the mass re-exports.
+
+# Background
+
+This proposal came out of the desire to attempt to isolate the processing
+pipeline out of the majority of the Sentry codebase.  The end goal for that is
+to be able to perform important changes to the event processing pipeline in
+isolation of the rest of the code base to reduce the time spent in CI for
+important changes to it.
+
+As such all code related to the processing pipeline should be moved into a clear
+structure and have a largely independent test setup (think moving all of processing
+related logic to `sentry.services.processing` or `sentry_processing` for better
+enforcability).
```

Comment:
I think we should put thought into how services overlap & intersect as they will. Core concepts like Organizations and Projects are everywhere, and as we deepen connections between issues, performance, profiling and replays we create tension with the independence of each service.

In a previous role we had a 'service API' which was the public interface that other services inside the application could use. A similar approach has some documentation [as shopify](https://shopify.engineering/deconstructing-monolith-designing-software-maximizes-developer-productivity).

Reply:
Agreed. I would like to handle that as a separate independent proposal though.